op-node/rollup/derive/engine_update.go

```diff
@@ -128,45 +199,86 @@ func confirmPayload(
 	updateSafe bool,
 	agossip async.AsyncGossiper,
 	sequencerConductor conductor.SequencerConductor,
-) (out *eth.ExecutionPayloadEnvelope, errTyp BlockInsertionErrType, err error) {
-	var envelope *eth.ExecutionPayloadEnvelope
+	builderClient builder.PayloadBuilder,
+	l2head eth.L2BlockRef,
+	metrics Metrics,
+) (*eth.ExecutionPayloadEnvelope, BlockInsertionErrType, error) {
+	var engineEnvelope *eth.ExecutionPayloadEnvelope
+	var builderPayload *PayloadRequestResult
+	var err error
 	// if the payload is available from the async gossiper, it means it was not yet imported, so we reuse it
 	if cached := agossip.Get(); cached != nil {
-		envelope = cached
+		engineEnvelope = cached
 		// log a limited amount of information about the reused payload, more detailed logging happens later down
 		log.Debug("found uninserted payload from async gossiper, reusing it and bypassing engine",
-			"hash", envelope.ExecutionPayload.BlockHash,
-			"number", uint64(envelope.ExecutionPayload.BlockNumber),
-			"parent", envelope.ExecutionPayload.ParentHash,
-			"txs", len(envelope.ExecutionPayload.Transactions))
+			"hash", engineEnvelope.ExecutionPayload.BlockHash,
+			"number", uint64(engineEnvelope.ExecutionPayload.BlockNumber),
+			"parent", engineEnvelope.ExecutionPayload.ParentHash,
+			"txs", len(engineEnvelope.ExecutionPayload.Transactions))
 	} else {
-		envelope, err = eng.GetPayload(ctx, payloadInfo)
+		engineEnvelope, builderPayload, err = getPayloadWithBuilderPayload(ctx, log, eng, payloadInfo, l2head, builderClient, metrics)
+		if err != nil {
+			// even if it is an input-error (unknown payload ID), it is temporary, since we will re-attempt the full payload building, not just the retrieval of the payload.
+			return nil, BlockInsertTemporaryErr, fmt.Errorf("failed to get execution payload from engine: %w", err)
+		}
 	}
-	if err != nil {
-		// even if it is an input-error (unknown payload ID), it is temporary, since we will re-attempt the full payload building, not just the retrieval of the payload.
-		return nil, BlockInsertTemporaryErr, fmt.Errorf("failed to get execution payload: %w", err)
+
+	engineBid := WeiToGwei(engineEnvelope.BlockValue)
+	if builderPayload != nil && builderPayload.success {
+		builderBid := WeiToGwei(builderPayload.envelope.BlockValue)
+		// TODO: boost factor for payload value
+		if builderBid >= engineBid {
```

Comment:
the comparison should not be using the converted Gwei value, but the raw Wei value. There is precision lost.

Reply:
I was refering to the prysm implementation but yeah why not more precision
https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix.go#L123

Reply:
I think they have less precision because they add a "boost" factor